File: anilist_completed_diff_to_csv.mjs
// anilist_completed_diff_to_csv.mjs
import { writeFile } from "node:fs/promises";

const ENDPOINT = "https://graphql.anilist.co";
const QUERY = /* GraphQL */ `
query CompareCompleted($user1: String!, $user2: String!) {
  u1: MediaListCollection(userName: $user1, type: ANIME, status_in: [COMPLETED]) {
    lists {
      entries {
        media {
          id
          title { romaji english native }
          siteUrl
        }
      }
    }
  }
  u2: MediaListCollection(userName: $user2, type: ANIME, status_in: [COMPLETED]) {
    lists {
      entries {
        media {
          id
          title { romaji english native }
          siteUrl
        }
      }
    }
  }
}
`;

function csvEscape(val) {
  if (val === null || val === undefined) return "";
  const s = String(val);
  // ダブルクォートと改行を考慮してRFC4180風に
  if (/[",\n]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

function toCSV(rows, header) {
  const head = header.join(",");
  const body = rows
    .map((r) => header.map((h) => csvEscape(r[h])).join(","))
    .join("\n");
  return `${head}\n${body}\n`;
}

function collectMedia(coll) {
  if (!coll?.lists) return [];
  const medias = [];
  for (const lst of coll.lists) {
    for (const e of lst.entries ?? []) {
      if (e?.media) medias.push(e.media);
    }
  }
  // idでユニーク化
  const uniq = new Map();
  for (const m of medias) uniq.set(m.id, m);
  return [...uniq.values()];
}

async function fetchCompleted(user1, user2) {
  const token = process.env.ANILIST_TOKEN?.trim();
  const headers = {
    "Content-Type": "application/json",
    Accept: "application/json",
  };
  if (token) headers.Authorization = `Bearer ${token}`;

  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers,
    body: JSON.stringify({ query: QUERY, variables: { user1, user2 } }),
    // Node の fetch には timeout option がないため AbortController を使う場合は別途実装
  });

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`AniList API error: ${res.status} ${res.statusText}\n${text}`);
  }

  const data = await res.json();
  if (data.errors) {
    throw new Error(`AniList GraphQL errors:\n${JSON.stringify(data.errors, null, 2)}`);
  }
  return data.data;
}

async function main() {
  const user1 = process.argv[2] || "tayo6";
  const user2 = process.argv[3] || "namonaki";
  const outPath =
    process.argv[4] || `anilist_diff_${user1}_vs_${user2}.csv`;

  const { u1, u2 } = await fetchCompleted(user1, user2);
  const u1Media = collectMedia(u1);
  const u2Media = collectMedia(u2);

  const set1 = new Set(u1Media.map((m) => m.id));
  const set2 = new Set(u2Media.map((m) => m.id));

  const idx1 = new Map(u1Media.map((m) => [m.id, m]));
  const idx2 = new Map(u2Media.map((m) => [m.id, m]));

  const onlyInUser1 = [...set1].filter((id) => !set2.has(id)).map((id) => idx1.get(id));
  const onlyInUser2 = [...set2].filter((id) => !set1.has(id)).map((id) => idx2.get(id));

  // 出力行作成
  const rows = [];
  for (const m of onlyInUser1.sort((a, b) => a.id - b.id)) {
    rows.push({
      side: `only_in_${user1}`,
      media_id: m.id,
      title_romaji: m.title?.romaji ?? "",
      title_english: m.title?.english ?? "",
      title_native: m.title?.native ?? "",
      site_url: m.siteUrl ?? "",
    });
  }
  for (const m of onlyInUser2.sort((a, b) => a.id - b.id)) {
    rows.push({
      side: `only_in_${user2}`,
      media_id: m.id,
      title_romaji: m.title?.romaji ?? "",
      title_english: m.title?.english ?? "",
      title_native: m.title?.native ?? "",
      site_url: m.siteUrl ?? "",
    });
  }

  const header = ["side", "media_id", "title_romaji", "title_english", "title_native", "site_url"];
  const csv = toCSV(rows, header);
  await writeFile(outPath, csv, "utf8");
  console.log(`Done: ${outPath} (${rows.length} rows)`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: anilist_completed_diff_to_csv.py
# anilist_completed_diff_to_csv.py
import csv
import json
import sys
import requests

ENDPOINT = "https://graphql.anilist.co"
QUERY = """
query CompareCompleted($user1: String!, $user2: String!) {
  u1: MediaListCollection(userName: $user1, type: ANIME, status_in: [COMPLETED]) {
    lists {
      entries {
        media {
          id
          title { romaji english native }
          siteUrl
        }
      }
    }
  }
  u2: MediaListCollection(userName: $user2, type: ANIME, status_in: [COMPLETED]) {
    lists {
      entries {
        media {
          id
          title { romaji english native }
          siteUrl
        }
      }
    }
  }
}
"""

def collect_media(coll):
    """MediaListCollection → [media dict ...]"""
    if not coll or "lists" not in coll or coll["lists"] is None:
        return []
    medias = []
    for lst in coll["lists"]:
        for e in (lst.get("entries") or []):
            m = e.get("media")
            if m:
                medias.append(m)
    # 万一の重複に備えてidでユニーク化
    uniq = {}
    for m in medias:
        uniq[m["id"]] = m
    return list(uniq.values())

def main(user1="tayo6", user2="namonaki", out_path=None):
    if out_path is None:
        out_path = f"anilist_diff_{user1}_vs_{user2}.csv"

    resp = requests.post(
        ENDPOINT,
        json={"query": QUERY, "variables": {"user1": user1, "user2": user2}},
        headers={"Content-Type": "application/json", "Accept": "application/json"},
        timeout=30,
    )
    resp.raise_for_status()
    data = resp.json()
    if "errors" in data:
        raise RuntimeError(json.dumps(data["errors"], ensure_ascii=False, indent=2))

    u1_media = collect_media(data["data"].get("u1"))
    u2_media = collect_media(data["data"].get("u2"))

    set1 = {m["id"] for m in u1_media}
    set2 = {m["id"] for m in u2_media}

    only_in_user1_ids = set1 - set2
    only_in_user2_ids = set2 - set1

    # id→media の辞書
    idx1 = {m["id"]: m for m in u1_media}
    idx2 = {m["id"]: m for m in u2_media}

    rows = []

    for mid in sorted(only_in_user1_ids):
        m = idx1[mid]
        rows.append({
            "side": "only_in_tayo6",
            "media_id": m["id"],
            "title_romaji": (m["title"]["romaji"] or "").strip() if m.get("title") else "",
            "title_english": (m["title"]["english"] or "").strip() if m.get("title") else "",
            "title_native": (m["title"]["native"] or "").strip() if m.get("title") else "",
            "site_url": m.get("siteUrl", "")
        })

    for mid in sorted(only_in_user2_ids):
        m = idx2[mid]
        rows.append({
            "side": "only_in_namonaki",
            "media_id": m["id"],
            "title_romaji": (m["title"]["romaji"] or "").strip() if m.get("title") else "",
            "title_english": (m["title"]["english"] or "").strip() if m.get("title") else "",
            "title_native": (m["title"]["native"] or "").strip() if m.get("title") else "",
            "site_url": m.get("siteUrl", "")
        })

    # CSVへ出力
    fieldnames = ["side", "media_id", "title_romaji", "title_english", "title_native", "site_url"]
    with open(out_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)

    print(f"Done: {out_path}  ({len(rows)} rows)")

if __name__ == "__main__":
    # 使い方: python script.py [user1] [user2] [out_path]
    u1 = sys.argv[1] if len(sys.argv) > 1 else "tayo6"
    u2 = sys.argv[2] if len(sys.argv) > 2 else "namonaki"
    outp = sys.argv[3] if len(sys.argv) > 3 else None
    main(u1, u2, outp)
